tools/clang/unittests/HLSLExec/LongVectors.cpp

```diff
@@ -692,608 +505,989 @@ void fillShaderBufferFromLongVectorData(std::vector<BYTE> &ShaderBuffer,
+
+#define DISPATCH(TYPE, FUNC)                                                   \
+  if (Config.DataType == DataTypeName<TYPE>())                                 \
+  return dispatchTernaryMathOpTest(Config, OpType, VectorSize,                 \
+                                   TernaryMathOps::FUNC<TYPE>)
 
   switch (OpType) {
   case TernaryMathOpType_Fma:
+    DISPATCH(double, Fma);
+    break;
+
   case TernaryMathOpType_Mad:
+    DISPATCH(HLSLHalf_t, Mad);
+    DISPATCH(float, Mad);
+    DISPATCH(double, Mad);
+    DISPATCH(int16_t, Mad);
+    DISPATCH(int32_t, Mad);
+    DISPATCH(int64_t, Mad);
+    DISPATCH(uint16_t, Mad);
+    DISPATCH(uint32_t, Mad);
+    DISPATCH(uint64_t, Mad);
+    break;
+
   case TernaryMathOpType_SmoothStep:
+    DISPATCH(HLSLHalf_t, SmoothStep);
+    DISPATCH(float, SmoothStep);
+    break;
+
+  case TernaryMathOpType_EnumValueCount:
     break;
-  default:
-    LOG_ERROR_FMT_THROW(L"Invalid TernaryMathOpType: %ls", OpTypeName.c_str());
   }
 
-  auto ComputeFunc = [this](const T &A, const T &B, const T &C) {
-    return this->computeExpectedValue(A, B, C);
-  };
-  InitTernaryOpValueComputer<T>(ComputeFunc);
+  LOG_ERROR_FMT_THROW(L"DataType '%s' not supported for TernaryMathOpType '%s'",
+                      (const wchar_t *)Config.DataType,
+                      (const wchar_t *)Config.OpTypeEnum);
+}
+
+//
+// dispatchTest
+//
+
+template <typename OP_TYPE> OP_TYPE GetOpType(const wchar_t *OpTypeString);
+
+template <> TrigonometricOpType GetOpType(const wchar_t *OpTypeString) {
+  return getTrigonometricOpType(OpTypeString).OpType;
+}
+
+template <> UnaryOpType GetOpType(const wchar_t *OpTypeString) {
+  return getUnaryOpType(OpTypeString).OpType;
+}
+
+template <> AsTypeOpType GetOpType(const wchar_t *OpTypeString) {
+  return getAsTypeOpType(OpTypeString).OpType;
+}
+
+template <> UnaryMathOpType GetOpType(const wchar_t *OpTypeString) {
+  return getUnaryMathOpType(OpTypeString).OpType;
+}
+
+template <> BinaryMathOpType GetOpType(const wchar_t *OpTypeString) {
+  return getBinaryMathOpType(OpTypeString).OpType;
+}
+
+template <> TernaryMathOpType GetOpType(const wchar_t *OpTypeString) {
+  return getTernaryMathOpType(OpTypeString).OpType;
+}
+
+template <typename OP_TYPE> void dispatchTest(const TestConfig &Config) {
+  OP_TYPE OpType = GetOpType<OP_TYPE>(Config.OpTypeEnum);
+
+  std::vector<size_t> InputVectorSizes;
+  if (Config.LongVectorInputSize)
+    InputVectorSizes.push_back(Config.LongVectorInputSize);
+  else
+    InputVectorSizes = {3, 4, 5, 16, 17, 35, 100, 256, 1024};
+
+  for (size_t VectorSize : InputVectorSizes) {
+    dispatchTest(Config, OpType, VectorSize);
+  }
+}
+
+// TAEF test entry points
+
+TEST_F(OpTest, trigonometricOpTest) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+
+  auto Config = TestConfig::Create(VerboseLogging);
+  if (Config)
+    dispatchTest<TrigonometricOpType>(*Config);
```

Comment:
I don't feel strongly about it but I've seen others suggest something like

```suggestion
  if (auto Config = TestConfig::Create(VerboseLogging))
    dispatchTest<TrigonometricOpType>(*Config);
```

Reply:
Oh thanks, this is the sort of thing I'd normally be calling out :)

Reply:
oh that's embarrassing.